### DISCUSSION TITLE
Hard Constraints

### DISCUSSION THREAD

**Discussion by 101AlexMartin** [lululxvi/deepxde#1382] [Q&A] [answered]
First of all thanks for the library, it's really helpful. I was wondering if there is any way to impose hard constraints in the solution provided by the network. For instance, let's say I'm asking it to solve for `[v_x, v_y, v_z]` on a square domain, and I need that all the solution `v_z` are positive, since negative `v_z` would mean an unfeasible solution in my invented problem. You can always add an additional term to the residuals of the pdes in the loss function, such as `max(-v_z,0)` and play with the weights, but this is still a soft constraint, does not ensure `v_z` will be positive definite.

**Comment by xuelanghanbao** [comment, marked as answer]
You can read [this section of the doc](https://deepxde.readthedocs.io/en/latest/demos/pinn_forward/poisson.dirichlet.1d.exactbc.html#poisson-equation-in-1d-with-hard-boundary-conditions), it goes through the implementation of an output transform for the net and which can be seen as a hard constraint. In your example, you can add the following output transform:
```
def  output_transform(x, y):
    return np.exp(y)

net.apply_output_transform(output_transform)
```
The above code uses $e ^ x$ to map the output to $(0,+\infty)$. 
However, it is worth noting that using output transform is equivalent to decomposing the results. If the true solution does not meet this decomposition, it may make it more difficult for the network to converge
